_posts/2019-03-18-legacy-rails-apps-silently-judging-you.md

```diff
@@ -0,0 +1,236 @@
+`./bin/setup` script? Do they run?
+
+Setting up a new development environment should be as simple as:
+
+```
+./bin/setup`
+```
+
+Or:
+
+```
+docker-compose up
+```
+
+Or:
+
+```
+vagrant up
+```
+
+You **get** the point. It is a **red flag** when you need to pair with someone to get
+your development environment set up. You should count the times you run into
+unexpected issues. This will give you a good idea of the complexity of *all*
+environments.
+
+## Build Judgment
+
+Is there a set of spec files? Does the build work in your environment? After you
+have set up your local environment, you should be able to run:
+
+```
+rake
+```
+
+I'm used to projects and gems using `test` or `spec` as the default Rake task,
+which has been the default task for Rails for a while: [https://github.com/rails/rails/blob/5-2-stable/Rakefile#L21-L22](https://github.com/rails/rails/blob/5-2-stable/Rakefile#L21-L22)
+
+```ruby
+# Rakefile
+desc "Run all tests by default"
+task default: %w(test test:isolated)
+```
+
+It is a **red flag** when you run into unexpected issues when running the build.
+If some tests failed, why did they fail? How many random failures did you get? If
+you found many issues, why didn't `./bin/setup` take care of it for you beforehand?
+
+## Coverage Judgment
+
+Let's say your application is so well maintained that all previous steps worked
+flawlessly. `bundle outdated` returned no outdated dependencies; static code
+analysis reported an A+; you run `./bin/setup` to install everything you needed;
+and the `rake` showed you a passing build. That is great!
+
+How well covered is it? Does it cover more than 80% of your code? Does it cover
+100% of the [critical path]()?
+
+You can quickly find out with `SimpleCov`. Simply add this to your `Gemfile`:
+
+```ruby
+# Gemfile
+gem 'simplecov', require: false, group: :test
+```
+
+Then add this to your test helper file:
+
+```ruby
+if ENV['COVERAGE'] == 'true'
+  require 'simplecov'
+  SimpleCov.start 'rails'
+end
+```
+
+Then `bundle install` and run the build. You should get a report like this one:
+
+<img src="/blog/assets/images/simple-cov-report.png" alt="SimpleCov Report">
+
+## 12 Factor Judgment
+
+Yes, yes, we all know we **should be** following the [twelve-factor methodology](https://fastruby.io/twelve-factor) but some people _missed the memo_.
+
+How many values are hardcoded in the configuration files? Are there more than 3
+environments? How are environment variables managed? Are they using a
+[`config/secrets.yml`](https://guides.rubyonrails.org/4_1_release_notes.html#config-secrets-yml) file? Is `config/database.yml` checked in to the repository? If so,
+is it using environment variables? Are all dependencies clearly declared? How
+are they tracking exceptions in production? What does the time between deployments
+gap look like?
+
+## MVC Judgment
+
+One of the best selling points of Rails is convention over configuration. You
```

Comment:
@etagwerker You can link `convention over configuration` to https://rubyonrails.org/doctrine/#convention-over-configuration